blob/azureblob/azureblob.go

```diff
@@ -661,6 +661,7 @@ func (b *bucket) SignedURL(ctx context.Context, key string, opts *driver.SignedU
 		ContainerName: b.name,
 		BlobName:      srcBlobParts.BlobName,
 		Permissions:   perms.String(),
+		ContentType:   opts.ContentType,
```

Comment:
azure tests are failing:

```
    --- FAIL: TestConformance/TestSignedURL (0.67s)
        setup.go:79: Recording into golden file testdata/TestConformance/TestSignedURL.replay
        drivertest.go:2120: PUT with PUT URL (content-type="application/octet-stream") got status code 201, want 2xx? false
        drivertest.go:2122: 
        drivertest.go:2120: PUT with PUT URL (content-type="") got status code 201, want 2xx? false
        drivertest.go:2122: 
```

Reply:
What do you mean by "they don't really conform to our contract for PUT URLs"? Is there something else other than the content-type?

We could just relax the contract, e.g.:

```
        // ContentType specifies the Content-Type HTTP header the user agent is
        // permitted to use in the PUT request. It must match exactly. Empty means
        // the Content-Type header is not permitted.
```
--

```
        // ContentType specifies the Content-Type HTTP header the user agent is
        // expected to use in the PUT request. Some providers enforce it, others do not.
        // When enforced, it must match exactly, and an empty string means the
        // Content-Type header is not permitted.
```

or something like that.

Reply:
I'm referring to the issue you identified originally in #2630 about Azure requiring an additional `x-ms-blob-type` header which the others don't.

Reply:
Ah yes, I forgot about that.

Let's:
-- return a `NotImplemented` error for `azblob` if the `ContentType` field is set
-- check for `NotImplemented` in the conformance test to skip the test
-- update the comment about `PUT` not being fully conformant at the top of `azblob.go` to also note that `opts.ContentType` can't be set.

WDYT?

Reply:
It doesn't really fit the spirit of the contract: if the `ContentType` field isn't set, that's supposed to indicate that a `Content-Type` header **can't** be set on the `PUT` request (this is the behavior that S3 and GCS enforce). Thus, my recommendation to just forbid `PUT` signed URLs on Azure, because I think it's too far away from the contract the others align with.

Reply:
That's a backward incompatible change though; there may be users using `PUT` signed URLs today.

Reply:
Sure, but this is functionality that was only introduced in this last release IIUC.

The reason I'm pushing is because it significantly weakens what guarantees can be made of the `ContentType` field. Since Azure doesn't enforce the `ContentType` field's behavior at all, then the Azure implementation has to go out of its way to say "if you use `PUT` signed URLs at all, it's non-compliant and behaves in a totally different way than the other implementations".

If you really feel strongly that we should keep the behavior, I'll drop it to move this PR forward, since I'd very much like to see this fix for S3 and GCS. I bumped into the problem the instant I tried to use the new `PUT` URLs on GCS and the failure modes are really hard to debug from the GCS side. I fear other users will encounter the same issues and get frustrated with the CDK.

Reply:
What do you think about making `ContentType` be a `*string` instead of a `string`, where:

* `nil` (the default) means "no assertion about ContentType"
* a non-`nil` value means an assertion ("" asserts no ContentType header)
* There's a comment that says that not all providers implement verification of non-nil `ContentType` values (which would be true for Azure).

Reply:
Alright, I've added a new field `EnforceAbsentContentType`, WDYT? It's spiritually pretty close to what you suggested, but it has two advantages:

1. It doesn't require the API surface awkwardness of `new(string)` or a helper function.
2. It makes it easier to communicate that a `(ContentType, EnforceAbsentContentType)` of `("", false)` may still perform enforcement, depending on implementation (i.e. GCS and S3). `("", true)` will always perform enforcement or will raise an error (Azure). Most implementations will never need to care about the distinction and can just read from `ContentType`.

Thanks for the patience here. Work has been busy, and I wanted to wait until I had adequate time to ponder the correct API.